File: bspline_fitting/Cpp/src/trainer.cpp
#include "trainer.h"

using namespace pybind11::literals;

Trainer::Trainer(const std::string root_path, const int degree_u,
                 const int degree_v, const int size_u, const int size_v,
                 const int sample_num_u, const int sample_num_v,
                 const float start_u, const float start_v, const float stop_u,
                 const float stop_v, const std::string device,
                 const int warm_epoch_step_num, const int warm_epoch_num,
                 const int finetune_step_num, const float lr,
                 const float weight_decay, const float factor,
                 const int patience, const float min_lr,
                 const std::string save_result_folder_path,
                 const std::string save_log_folder_path) {
  py::gil_scoped_acquire acquire;

  py::object sys = py::module_::import("sys");

  sys.attr("path").attr("append")(root_path);

  py::object Trainer = py::module_::import("bspline_fitting.Module.trainer");

  trainer_ = Trainer.attr("Trainer")(
      "degree_u"_a = degree_u, "degree_v"_a = degree_v, "size_u"_a = size_u,
      "size_v"_a = size_v, "sample_num_u"_a = sample_num_u,
      "sample_num_v"_a = sample_num_v, "start_u"_a = start_u,
      "start_v"_a = start_v, "stop_u"_a = stop_u, "stop_v"_a = stop_v,
      "device"_a = device, "warm_epoch_step_num"_a = warm_epoch_step_num,
      "warm_epoch_num"_a = warm_epoch_num,
      "finetune_step_num"_a = finetune_step_num, "lr"_a = lr,
      "weight_decay"_a = weight_decay, "factor"_a = factor,
      "patience"_a = patience, "min_lr"_a = min_lr,
      "save_result_folder_path"_a = save_result_folder_path,
      "save_log_folder_path"_a = save_log_folder_path);

  return;
}

Trainer::~Trainer() {}

const bool Trainer::toBSplineSurface(const std::vector<float> &sample_points) {
  py::gil_scoped_acquire acquire;

  py::list sample_point_list;
  for (int i = 0; i < sample_points.size(); ++i) {
    sample_point_list.append(sample_points[i]);
  }

  py::print("start autoTrainBSplineSurface");
  const bool success =
      trainer_
          .attr("autoTrainBSplineSurface")("gt_points"_a = sample_point_list)
          .cast<bool>();

  return success;
}

const std::vector<float> Trainer::getKNotsU() {
  py::list list = trainer_.attr("bspline_surface").attr("toKNotsUList")();

  std::vector<float> knots_u;
  knots_u.reserve(list.size());

  for (int i = 0; i < list.size(); ++i) {
    knots_u.emplace_back(list[i].cast<float>());
  }

  return knots_u;
}

const std::vector<float> Trainer::getKNotsV() {
  py::list list = trainer_.attr("bspline_surface").attr("toKNotsVList")();

  std::vector<float> knots_v;
  knots_v.reserve(list.size());

  for (int i = 0; i < list.size(); ++i) {
    knots_v.emplace_back(list[i].cast<float>());
  }

  return knots_v;
}
const std::vector<float> Trainer::getCtrlPts() {
  py::list list = trainer_.attr("bspline_surface").attr("toCtrlPtsList")();

  std::vector<float> ctrl_pts;
  ctrl_pts.reserve(list.size());

  for (int i = 0; i < list.size(); ++i) {
    ctrl_pts.emplace_back(list[i].cast<float>());
  }

  return ctrl_pts;
}

File: pmodule/Module/trainer.py
import os
import torch
import numpy as np
import open3d as o3d
from tqdm import tqdm
from typing import Union
from copy import deepcopy
from torch.optim import Optimizer, AdamW
from torch.optim.lr_scheduler import (
    LRScheduler,
    CosineAnnealingWarmRestarts,
    ReduceLROnPlateau,
)

from bspline_fitting.Config.constant import EPSILON
from bspline_fitting.Loss.chamfer_distance import chamferDistance
from bspline_fitting.Method.pcd import getPointCloud
from bspline_fitting.Method.time import getCurrentTime
from bspline_fitting.Method.fitting import approximate_surface
from bspline_fitting.Model.bspline_surface import BSplineSurface
from bspline_fitting.Module.logger import Logger
from bspline_fitting.Module.o3d_viewer import O3DViewer


class Trainer(object):
    def __init__(
        self,
        degree_u: int = 3,
        degree_v: int = 3,
        size_u: int = 5,
        size_v: int = 7,
        sample_num_u: int = 20,
        sample_num_v: int = 20,
        start_u: float = 0.0,
        start_v: float = 0.0,
        stop_u: float = 1.0,
        stop_v: float = 1.0,
        idx_dtype=torch.int64,
        dtype=torch.float64,
        device: str = "cpu",
        warm_epoch_step_num: int = 20,
        warm_epoch_num: int = 10,
        finetune_step_num: int = 400,
        lr: float = 1e-2,
        weight_decay: float = 1e-4,
        factor: float = 0.9,
        patience: int = 1,
        min_lr: float = 1e-4,
        render: bool = False,
        render_freq: int = 1,
        render_init_only: bool = False,
        save_result_folder_path: Union[str, None] = None,
        save_log_folder_path: Union[str, None] = None,
    ) -> None:
        return

    def trainBSplineSurface(
        self,
        optimizer: Optimizer,
        scheduler: LRScheduler,
        gt_points: torch.Tensor,
    ) -> bool:
        self.bspline_surface.setGradState(True)

        train_step_num = self.toTrainStepNum(scheduler)
        final_step = self.step + train_step_num

        print("[INFO][BSplineSurfaceModelOp::train]")
        print("\t start training ...")
        pbar = tqdm(total=final_step)
        pbar.update(self.step)
        while self.step < final_step:
            loss_dict = self.trainStep(
                optimizer,
                gt_points,
            )

            assert isinstance(loss_dict, dict)
            if self.logger.isValid():
                for key, item in loss_dict.items():
                    self.logger.addScalar("Train/" + key, item, self.step)
                self.logger.addScalar("Train/lr", self.getLr(optimizer), self.step)

            self.updateBestParams(loss_dict["loss"])

            pbar.set_description(
                "LOSS %.6f LR %.4f"
                % (
                    loss_dict["loss"],
                    self.getLr(optimizer) / self.lr,
                )
            )

            self.autoSaveBSplineSurface("train")

            if self.checkStop(optimizer, scheduler, loss_dict):
                break

            self.step += 1
            pbar.update(1)

        return True

    def autoTrainBSplineSurface(
        self,
        gt_points: Union[np.ndarray, list, tuple],
    ) -> bool:
        self.bspline_surface.ctrlpts.data = self.bspline_surface.ctrlpts.data * scale + torch.from_numpy(center)

        return True
